### DISCUSSION TITLE
[Meta] Experience with Upgrade to PHP 8.0

### DISCUSSION THREAD

**Discussion by TomasVotruba** [rectorphp/rector#6617] [General]
This posts mentions many small details that Rector probably misses now.

It would be great to cover them to make PHP 8.0 set more complete: 
https://medium.com/oro-development/upgrade-to-php-8-64f770ae4479

It's a great first issue candidate, as changes are very small. Pick the one you like and try to improve existing rule and create a new one :+1:

**Comment by marekroman** [comment]
Hello, this is probably a wrong place to post this, but the biggest issue with upgrading 7.4 to 8.0 is: 

1) the string to number comparison behavior change (zero == empty string in php 7.4, but not in php 8.0) and 

2) internal function warnings now throw TypeError (e.g. attempting to use count/in_array/... with null or non-countable/non-array params). 

While the problem number 2) is annoying, it is fixable, since the script typically crashes in a very obvious way a can therefore be easily fixed (besides, this type of bad code using non-array data with array functions is relatively rare). But the problem 1) is a massive issue (dare I say it's s disaster for anyone attempting to port any bigger, say 100k+ line, codebase with lots of legacy code). 

It is a massive issue because: 
a) that kind of code is very common (expecting that 0 == '' is true), 
b) the compared variables typically come from a POSTed form, session or a database during runtime
c) the resulting change in behavior fundamentally changes the script execution path while producing no errors

Due to the point b) above, I don't think it's safe to upgrade to php8 using any kind of automated tool and thanks to c) this is the single most backwards incompatible change PHP has ever done. The PHP team just messed that one up.

**Comment by TomasVotruba** [comment]
Hi, thanks for sharing your experience :+1: 

Non-constant types are hard to deal with, mainly the weak ones. Personally I would upgrade to strict comparison first, deal with the bugs and only then play with PHP upgrade.
